libqpdf/QPDF_optimization.cc

```diff
@@ -277,78 +285,70 @@ QPDF::pushInheritedAttributesToPageInternal(
-            return;
+        if (cur.oh.isIndirect()) {
+            QPDFObjGen og(cur.oh.getObjGen());
+            if (!visited.add(og)) {
+                QTC::TC("qpdf", "QPDF opt loop detected");
+                continue;
+            }
+            m->obj_user_to_objects[cur.ou].insert(og);
+            m->object_to_obj_users[og].insert(cur.ou);
         }
-        m->obj_user_to_objects[ou].insert(og);
-        m->object_to_obj_users[og].insert(ou);
-    }
 
-    if (oh.isArray()) {
-        int n = oh.getArrayNItems();
-        for (int i = 0; i < n; ++i) {
-            updateObjectMapsInternal(
-                ou, oh.getArrayItem(i), skip_stream_parameters, visited, false);
-        }
-    } else if (oh.isDictionary() || oh.isStream()) {
-        QPDFObjectHandle dict = oh;
-        bool is_stream = oh.isStream();
-        int ssp = 0;
-        if (is_stream) {
-            dict = oh.getDict();
-            if (skip_stream_parameters) {
-                ssp = skip_stream_parameters(oh);
+        if (cur.oh.isArray()) {
+            int n = cur.oh.getArrayNItems();
+            for (int i = 0; i < n; ++i) {
+                pending.emplace_front(cur.ou, cur.oh.getArrayItem(i), false);
+            }
+        } else if (cur.oh.isDictionary() || cur.oh.isStream()) {
+            QPDFObjectHandle dict = cur.oh;
+            bool is_stream = cur.oh.isStream();
+            int ssp = 0;
+            if (is_stream) {
+                dict = cur.oh.getDict();
+                if (skip_stream_parameters) {
+                    ssp = skip_stream_parameters(cur.oh);
+                }
             }
-        }
 
-        for (auto const& key: dict.getKeys()) {
-            if (is_page_node && (key == "/Thumb")) {
-                // Traverse page thumbnail dictionaries as a special case.
-                updateObjectMapsInternal(
-                    ObjUser(ObjUser::ou_thumb, ou.pageno),
-                    dict.getKey(key),
-                    skip_stream_parameters,
-                    visited,
-                    false);
-            } else if (is_page_node && (key == "/Parent")) {
-                // Don't traverse back up the page tree
-            } else if (
-                ((ssp >= 1) && (key == "/Length")) ||
-                ((ssp >= 2) && ((key == "/Filter") || (key == "/DecodeParms")))) {
-                // Don't traverse into stream parameters that we are not going to write.
-            } else {
-                updateObjectMapsInternal(
-                    ou, dict.getKey(key), skip_stream_parameters, visited, false);
+            for (auto const& key: dict.getKeys()) {
+                if (is_page_node && (key == "/Thumb")) {
+                    // Traverse page thumbnail dictionaries as a special case.
+                    auto ou = std::make_shared<ObjUser>(ObjUser::ou_thumb, cur.ou.pageno);
+                    pending.emplace_front(*ou, dict.getKey(key), false);
+                    heap.emplace_back(ou);
```

Comment:
... with 
```c++
heap.emplace_back(ObjUser::ou_thumb, cur.ou.pageno);
pending.emplace_back(heap.back(());
```
We may not even need a vector. A scalar ObjUser variable may be sufficient.

Reply:
I actually figure out how to do it without seeing your example, suggesting my brain is starting to extend around emplace_back. :-) I think I need a vector. I don't want to assume there isn't some invalid data that would make there be more than one of these.

Reply:
I don't need a vector. I do need `top` though, but I can't find your comment so maybe you realized and removed it.

Reply:
The ```if(is_page_node) ``` should ensure that this is a top-level key, preventing the possibility of any other `/Thumb` keys being processed.

Reply:
I can't spot why `top` is needed.

Reply:
> The `if(is_page_node) ` should ensure that this is a top-level key, preventing the possibility of any other `/Thumb` keys being processed.

Right, I explained that in my comment. I think `top` is needed because pending could become empty at any time, such as when traversing the last key of the dictionary.

Reply:
> Right, I explained that in my comment. 

I only saw your update after my comment.

>  I think top is needed because pending could become empty at any time

Good point. I think this needs a test case. Probably also for  copyForeignObject which could similarly get caught out.

Reply:
> I only saw your update after my comment.

We each did that to each other multiple times in this PR, which I guess shows we're thinking along the same path -- a good sign.



> Good point. I think this needs a test case. Probably also for copyForeignObject which could similarly get caught out.

Perhaps. We'd have to do it with a single-key dictionary to make it deterministic, and this would be a good case for a QTC coverage case since it's easy for the test to pass without exercising the boundary condition in the event of a logic change. This is the kind of thing that is what prevents me from doing too much optimization. With `top`, the intent is clear, and there is no need for tricky or subtle reasoning or fragility with edge cases, and there's no real performance cost other than the extra few bytes.